doc/wave.1
'\" t
.TH WAVE 1 "11 January 2000" "WAVE 6.3" "WFDB applications"
.SH NAME
wave \- waveform analyzer, viewer, and editor for the X Window System
.SH SYNOPSIS
\fBwave -r \fIrecord\fR[\fB+\fIrecord\fR ...] [ \fIoptions\fR ... ]
.SH DESCRIPTION
.PP
\fIwave\fR can be used to view the specified WFDB \fIrecord\fR or
records on any display controlled by an X11 server.  It includes
facilities for interactive annotation editing.  The keyboard and mouse
are used to control the display interactively.  First-time users
should read the \fIWAVE User's Guide\fR.  (One way to do this is by
pointing your Web browser to \fBfile:///usr/local/help/html/wug/wug.htm\fR,
or to \fBhttp://www.physionet.org/physiotools/wug/\fR for the latest version.)
If the \fIGuide\fR is unavailable, print a copy of the on-line manual; run
\fIwave\fR without any command-line arguments to get instructions for doing so.
Note that this operation does not require an X display.
.PP
If you specify more than one \fIrecord\fR, a separate \fIwave\fR process is
started for each record.  Note that all records to be opened must be listed in
a single command-line argument following \fB-r\fR, with \fB+\fR characters (not
spaces) between the record names.  See `Running two or more WAVE processes'
below.
.PP
Use the left mouse button to make selections, and the right mouse button to
open menus (indicated by triangular glyphs at the right end of some buttons).
See the \fIGuide\fR or the on-line manual for notes on annotation editing.
.PP
\fIOptions\fR are:
.TP
\fB-a \fIannotator\fR
Open the specified annotation file for the previously specified \fIrecord\fR
or records.
.TP
\fB-dpi \fIxx\fR[\fBx\fIyy\fR]
Calibrate \fIwave\fR for use with a display having a resolution of \fIxx\fR
(by \fIyy\fR) dots per inch.
.TP
\fB-f \fItime\fR
Open the record(s) beginning at the specified \fItime\fR.
.TP
\fB-g\fR
Use shades of grey only, even on a color monitor.
.TP
\fB-H\fR
Read the signal files in high-resolution mode (default: standard mode).
These modes are identical for ordinary records.  For multifrequency records,
the standard decimation of oversampled signals to the frame rate is suppressed
in high-resolution mode (rather, all other signals are resampled at the highest
sampling frequency).
.TP
\fB-m\fR
Use monochrome (usually black and white) only, even on a color or greyscale
monitor.  The line styles selected by the \fB-m\fR option may be easier to
distinguish on some greyscale monitors than the default shades of grey.
.TP
\fB-O\fR
Use overlay graphics for maximum speed and display quality if possible.  This
is the usual default if the X server supports a PseudoColor or GrayScale
visual.  This option exists only to force use of overlay graphics if a
different mode has been chosen as the default.
.TP
\fB-s \fIsignal\fR [\fIsignal ...\fR]
Initialize the signal list.  By default, the signal list includes all available
signals, in numerical order.
.TP
\fB-S\fR
Use the standard (shared) color palette, even if it is possible to modify
the palette.  Using this option conserves color resources if you have other
applications that use non-standard colors, at the expense of some speed in
redrawing the display.  The \fB-S\fR option may be used in combination with
the \fB-g\fR option if desired.
.TP
\fB-V\fIx\fR
Set display option \fIx\fR.  See `Display Options' below for details.
.PP
Note that \fIwave\fR queries the X server to determine the display
capabilities and resolution;  it is not necessary to use the \fB-g\fR,
\fB-m\fR, or \fB-S\fR options unless you wish to restrict \fIwave\fR's use of
the available capabilities.  Use the \fB-dpi\fR option to override the server's
default resolution if it is incorrect and cannot be changed otherwise (see
comments below under `Resources').
.PP
The system on which \fIwave\fR runs (the ``host'' system) need not be the
same as the system to which your keyboard, mouse and display are connected
(the ``local'' system), provided only that the host and local systems are on
the same network.  If you wish to run \fIwave\fR remotely, it is usually
necessary to grant permission for the host system to open windows on the
local system's display (generally, this is accomplished using \fIxhost\fR on
the local system;  see the documentation for your X server for details).
Use \fIrlogin\fR, \fItelnet\fR, or a similar method to log into the host
system from the local system, and set the \fBDISPLAY\fR environment variable
on the host system appropriately (if the local system runs UNIX, the value
of \fBDISPLAY\fR should be \fIlocal-hostname\fB:0.0\fR in most cases;  again,
consult your X server documentation).
.SH ENVIRONMENT
\fIwave\fR uses many environment variables;  they are listed in this section
roughly in order of importance.  Many of them need not be set at all, since
\fIwave\fR uses reasonable default values in most cases.  Those that are set
must be set on the host system.
.TP
\fBDISPLAY\fR
The name of the X server and display you are using (see above).  If not
set, \fIwave\fR will not be able to open windows on your display unless your
display is attached to the host system directly (not via network connection).
.TP
\fBWFDB\fR
The database path (see \fIsetwfdb\fR(1)).  If not set, \fIwave\fR can find
database files only in the current directory.  If you edit annotation files
and wish to reopen them later, be sure that the current directory (in
which \fIwave\fR writes any edited annotation files) is the first
directory in your database path.
.TP
\fBWFDBCAL\fR
The WFDB calibration file (see \fIsetwfdb\fR(1) and \fIwfdbcal\fR(5)).  If not
set, \fIwave\fR may not scale signals other than ECGs correctly.
.TP
\fBWAVEMENU\fR
The name of the analysis menu file (see below);  if not set,
\fIwave\fR uses \fIwavemenu\fR if it exists in the current directory, or
\fB$MENUDIR\fI/wavemenu.def\fR otherwise.
.TP
\fBSHELL\fR
The command interpreter used within the Analysis Commands window;  if not set,
\fIwave\fR uses \fI/bin/sh\fR (the Bourne shell).  Other shell-related
variables, such as \fBPATH\fR, are also significant when \fIwave\fR is running
commands within the Analysis Commands window.
.TP
\fBEDITOR\fR
The name of the text editor to be used for modifying the analysis menu
file and the log file.  If not set, \fIwave\fR uses \fItextedit\fR.
.TP
\fBPRINTER\fR
The name of a PostScript printer to be used for paper output;  if not set,
\fIwave\fR uses the default printer.
.TP
\fBPSPRINT\fR
The command used to print PostScript data from the standard input;  if not
set, \fIwave\fR uses `\fIlpr -P$PRINTER\fR'.
.TP
\fBTEXTPRINT\fR
The command used to print text from the standard input;  if not
set, \fIwave\fR uses `\fIlpr -P$PRINTER\fR'.
.TP
\fBANNTAB\fR
The name of a file that contains custom annotation definitions
(see `Resources', below, for details).  If not set, \fIwave\fR uses standard
annotation definitions only.
.PP
The environment variables below are not needed unless the \fIwave\fR binary
distribution, or XView, has been installed in non-standard directories:
.TP
\fBHELPPATH\fR
The path for XView spot help;  if not set, \fIwave\fR initializes it to
\fI/usr/lib/help\fR.  \fIwave\fR's own spot help is in
\fB$HELPDIR\fI/wave\fR, which is appended to the end of \fBHELPPATH\fR
by \fIwave\fR.
.TP
\fBHELPDIR\fR
The directory in which \fIwave\fR's help directory is located;  if not set,
\fIwave\fR uses \fI/usr/local/help\fR.
.TP
\fBMENUDIR\fR
The name of the directory that contains the default analysis menu
file;  if not set, \fIwave\fR uses \fI/usr/local/lib\fR.
.TP
\fBRESDIR\fR
The name of the directory in which system-wide default X11 resource files
are kept;  if not set, \fIwave\fR uses \fI/usr/lib/X11/app-defaults\fR.
\fBXUSERFILESEARCHPATH\fR, \fBXAPPLRESDIR\fR, and \fBXENVIRONMENT\fR are
also used, together with \fBHOME\fR and \fBUSER\fR, to locate resource files
(see \fIX\fR(1)).
.SH RESOURCES
.PP
You can control many aspects of \fIwave\fR's appearance and behavior by
setting its resources.  If you are not familiar with this concept, refer to an
introductory book on using the X Window System, such as Quercia and O'Reilly's
`X Window System User's Guide'.  Since \fIwave\fR is built using the XView
toolkit, all of the resources listed in \fIxview\fR(7) can be used with
\fIwave\fR.  In addition, the following \fIwave\fR-specific resources may
also be set:
.TP
\fBWave.AllowDottedLines\fR
This resource specifies if \fIwave\fR is allowed to render dotted
lines.  \fIwave\fR normally draws annotation marker bars as dotted
lines, and may use dotted lines for other display elements on
black-and-white displays for clarity.  Some X servers do not properly
render dotted lines, however;  if you observe irregular or missing
annotation marker bars, change the value of this resource from
\fBTrue\fR to \fBFalse\fR.
.TP
\fBWave.Anntab\fR
This resource specifies the name of a file that contains a table of
annotation definitions.  The environment variable \fBANNTAB\fR can also be used
to specify this filename;  the resource overrides the environment variable
if both are set.  The file contains one-line entries of the form
.br
        15 % Funny looking beat
.br
in which the first field specifies the (numeric) annotation code in the
range between 1 and \fBACMAX\fR inclusive (see
\fI/usr/include/wfdb/ecgcodes.h\fR for a list of predefined codes and for the
definition of \fBACMAX\fR);  the second field (`%' in the example) is a
mnemonic (used in annotation display and entry), and the remainder of the entry
is a description of the intended use of the annotation code (which appears next
to the mnemonic in the `Type' field and menu of `Annotation Template' windows).
Lines in the annotation table that begin with `#' are treated as comments and
ignored.  It is not necessary to specify an annotation table when editing an
existing annotation file unless previously undefined annotation types are to be
added to it during the editing process, although it is generally harmless
to do so.
.TP
\fBWave.Dpi\fR
This resource specifies the display resolution in dots per inch in the
form \fIMM\fBx\fINN\fR, where \fIMM\fR is the horizontal resolution and
\fINN\fR is the vertical resolution.  Normally, the resolution is known to the
X server, and it is unnecessary to specify this resource.  If your X server is
misinformed, \fIwave\fR's calibrated display scales will be incorrect;  the
best solution is to specify the resolution using a server option such as the
\fB-dpi\fR option supported by MIT's X11 servers, since this will solve
problems common to any other applications that require calibrated scales as
well.  Not all X11 servers support such an option, so this resource is
available as a work-around.  The command-line option \fB-dpi\fR  overrides the
resource if both are specified.
(If you don't know the resolution, use \fIxdpyinfo\fR(1) to determine what your
X server thinks it is.  Then run \fIwave\fR, enable the grid display, and
measure the grid squares with a ruler.  If they are larger than 5 mm, the
number of dots per inch returned by \fIxdpyinfo\fR is too large;  adjust the
\fBWave.Dpi\fR resource proportionally, and repeat the process until the
grid squares measure 5 mm in each direction.)
.TP
\fBWave.GraphicsMode\fR
This resource specifies the graphics mode used by \fIwave\fR;  it can be
overridden using the \fB-g\fR, \fB-m\fR, \fB-O\fR, or \fB-S\fR
options.  The legal values are \fB1\fR (monochrome mode), \fB2\fR
(overlay greyscale mode), \fB4\fR (shared color mode), \fB6\fR
(shared grey mode), and \fB8\fR (overlay color mode).
.TP
\fBWave.SignalWindow.\fR{\fBGrey\fR|\fBColor\fR}\fB.\fIElement\fR
These resources specify the colors to be used on greyscale or color
displays.  The `Color.*' resources are used only if the display is
color-capable and neither greyscale nor monochrome mode has been
specified.  The defaults are:
.TS
center;
l l l.
\fIElement\fB	Grey	Color\fR
\fBBackground\fR	white	white
\fBGrid\fR	grey75	grey90
\fBCursor\fR	grey50	orange red
\fBAnnotation\fR	grey25	yellow green
\fBSignal\fR	black	blue
.TE
.TP
\fBWave.SignalWindow.Mono.Background\fR
In monochrome mode, the background is normally white, and all other
display elements are normally black.  The reverse can be obtained by
setting this resource to \fBblack\fR.  (There is at least one server
for which this fails.)
.TP
\fBWave.Scope.\fR{\fBGrey\fR|\fBColor\fR}\fB.\fR{\fBForeground\fR|\fBBackground}\fR
These resources specify the colors to be used in the Scope window on greyscale
or color displays. The Foreground color is used for the waveform and the time
display; by default, it matches the color used for signals in the signal window
(see the previous item). Some X servers do not allow the background color of
the Scope window to be set, because of the color map animation and stippled
erasing techniques used. 
.TP
\fBWave.Scope.Mono.Background\fR
This resource can be used to invert the foreground and background of the Scope
window when WAVE is running in monochrome mode. This does not work for all X
servers. 
.TP
\fBWave.SignalWindow.{Height_mm|Width_mm}\fR
These resources specify the preferred dimensions (in millimeters) for the
signal window.  The defaults are 120 and 250 respectively.
.TP
\fBWave.SignalWindow.Font\fR
This resource specifies the font used to display annotations and time
marks in the signal window.  The default is \fIfixed\fR.
.TP
\fBWave.TextEditor\fR
This resource specifies the name of the text editor invoked by
\fIwave\fR to permit you to edit \fIwave\fR's log and analysis menu
files.  The default is \fItextedit\fR (the OpenLook visual editor).
You may override this resource by using the environment variable
\fBEDITOR\fR, which is also used by many other UNIX applications that
invoke editors.

.SS Display options
Initial values for the settings controlled from \fIwave\fR's View window can
be specified using either X resources or command-line options.  Once
suitable settings have been selected, use the `Save as new defaults'
button in \fIwave\fR's View window to record them in your \fI.Xdefaults\fR
file.  In this section, the X resource name is specified first, and
the command-line option follows.
.PP
By default, all of the display options in the first group are off
(\fBFalse\fR); set any of these X resources to \fBTrue\fR to enable
these options, or use the command-line options to do so.
.TP
\fBWave.View.Subtype\fR (\fB-Vs\fR)
Display annotation \fIsubtyp\fR fields.
.TP
\fBWave.View.Chan\fR (\fB-Vc\fR)
Display annotation \fIchan\fR fields.
.TP
\fBWave.View.Num\fR (\fB-Vn\fR)
Display annotation \fInum\fR fields.
.TP
\fBWave.View.Aux\fR (\fB-Va\fR)
Display annotation \fIaux\fR fields.
.TP
\fBWave.View.Markers\fR (\fB-Vm\fR)
Display annotation marker bars.
.TP
\fBWave.View.SignalNames\fR (\fB-VN\fR)
Display signal names along the left edge of the signal window.
.TP
\fBWave.View.Baselines\fR (\fB-Vb\fR)
Display baselines for any DC-coupled signals, and label the zero levels and the
units along the right edge of the signal window.
.TP
\fBWave.View.Level\fR (\fB-Vl\fR)
While the pointer is in the signal window and any mouse button is depressed,
track the intersections of the marker bar with the signals and draw
horizontal marker bars across the signal window at the levels of these
intersections.
.PP
The remaining resources and command-line display options correspond to
the menu buttons in \fIwave\fR's View window.  The value of each
resource, or the numeric argument that immediately follows the
command-line option, should match the position of the desired menu
choice, where the top item on each menu is in position 0, the one
below it is in position 1, etc.  For example, to set the initial
amplitude scale to 5 mm/mV (the item at position 2 in the `Amplitude
scale' menu), add \fB-Vv 2\fR to the command line, or
\fBWave.View.AmplitudeScale:2\fR to the X11 resource database.
.TP
\fBWave.View.TimeScale\fR (\fB-Vt\fR)
Set the time scale (0: 50 mm/min; 1: 125 mm/min; 2: 250 mm/min; 3: 500 mm/min;
4: 12.5 mm/sec; 5: 25 mm/sec (default); 6: 50 mm/sec; 7: 125 mm/sec;
8: 250 mm/sec).
.TP
\fBWave.View.AmplitudeScale\fR (\fB-Vv\fR)
Set the amplitude scale (0: 1 mm/mV; 1: 2.5 mm/mV; 2: 5 mm/mV; 3: 10 mm/mV
(default); 4: 20 mm/mV; 5: 40 mm/mV; 6: 100 mm/mV).
.TP
\fBWave.View.SignalMode\fR (\fB-VS\fR)
Set the choice on the `Draw' menu (0: all signals (default); 1: listed signals
only).
.TP
\fBWave.View.AnnotationMode\fR (\fB-VA\fR)
Set the choice on the `Show annotations' menu (0: centered (default); 1:
attached to signals; 2: as a signal).
.TP
\fBWave.View.TimeMode\fR (\fB-VT\fR)
Set the choice on the `Time display' menu (0: elapsed (default); 1: absolute;
2: in sample intervals).
.TP
\fBWave.View.GridMode\fR (\fB-VG\fR)
Set the choice on the `Grid' menu (0: none; 1: 0.2 s; 2: 0.5 mV; 3: 0.2s x 0.5 mV (default)).

.PP 
In addition to the usual ways of setting X resources, it is possible to set any
of those listed above, as well as any of the generic XView resources, by using
the \fB-xrm\fR or \fB-default\fR options on the command line when starting
\fIwave\fR. For example, you can set the background color of the signal window
using a command such as
.br
	\fBwave -r 100s -xrm Wave.SignalWindow.Color.Background:lightblue\fR

.SH RUNNING TWO OR MORE WAVE PROCESSES
.PP
By specifying two or more record names, separated by `\fB+\fR'
characters, in the command-line argument that follows `\fB-r\fR' (see
above), you may open separate WAVE signal windows (processes) for each
record.  These processes are almost completely independent: from any
signal window, you may navigate within the record, change display
settings, edit annotations, run external analysis programs, quit the
process, etc., without affecting any other signal windows.
.PP
For example, you may open two signal windows for the same record by:
.br
	\fBwave -r 100+100 -a atr\fR
.br
You can now move about the record freely in either window.  This facility
makes it easy to compare different segments of the record.
Note that whenever two or more windows are displaying the same set of
annotations, as in this case, only one should be editing the
annotations at any given time.
.PP
The window associated with the \fIlast\fR record named on the command
line has a special status: it is designated the master signal window,
and an extra button (labelled `Sync') appears at the top of this
window.  Clicking on this button causes all of the other signal
windows to be redrawn so that the times shown in their lower left
corners match that in the master signal window.  (Note, however, that
if you have quit a signal window from the middle of the list, any
signal windows from earlier in the list will no longer respond to sync
requests.)
.PP
By default, all command-line arguments apply to all signal windows.
You may specify an argument that is to apply to only one signal
window, however, by prefixing the argument with `\fB+\fIn\fB/\fR', where
\fIn\fR is the \fIsignal window number\fR.  (The first signal window,
corresponding to the first record named on the command line, is signal
window number 0; the next is number 1, etc.)
.PP
This facility has many applications.  For example, you may wish to open two
copies of the same record, with two different annotators:
.br
	\fBwave -r 100+100 -a +0/atr +1/qrs\fR
.br
In this case, record 100 is opened in two windows, with annotator
`atr' in window 0 and annotator `qrs' in window 1.  (The `\fB-a\fR'
option applies to both windows since it does not have a
`\fB+\fIn\fB/\fR' prefix.)
.PP
As another example, you may wish to discuss a record with colleagues
at other locations:
.br
	\fBwave -r 200+200+200 -a qrs +0/-display +0/atlantic.bigu.edu:0 \\\fR
.br
		\fB+1/-display +1/pacific.widget.com:0\fR
.br
Here, record 200 is opened in three windows.  Window 0 is opened on
display 0 of atlantic.bigu.edu, window 1 on display 0 of
pacific.widget.com, and window 2 (the master window) on the local
display.  (For this to work, your colleagues must first allow your
computer to open windows on their displays, typically using
\fIxhost\fR.  See xview(7) for information about the \fB-display\fR
option.  Notice that the `\fB+\fIn\fB/\fR' prefix must be attached to
both the `\fB-display\fR' option and to its argument in order to
apply both of these arguments to the same signal window.)
Your colleagues can freely move about the record, but you can direct
the discussion at any time by using the Sync button in your signal
window.  In a case such as this one, anyone can enable editing; you
should do so only after making sure that no one else has.  Once you
have saved your work (by selecting `Save' from the File menu), your
changes become visible to your colleagues if they reload the
annotations (by clicking on `Reload' from the Load window).
.PP
As a final example, the MIMIC Database includes both high-resolution
waveform records and medium-resolution (roughly 1 sample per second)
computed measurement records.  You may view both of these at the same
time using a command such as:
.br
	\fBwave -r 237+237n -a all\fR
.br
Typically, you will wish to view the high-resolution and low-resolution
data at different time scales.  Although \fIwave\fR attempts to choose
reasonable defaults, you can adjust the scales independently if you wish:
.br
	\fBwave -r 237+237n -a all +1/-Vt +1/2\fR
.PP
If you use \fIwavescript\fR or \fIwave-remote\fR to control the master
signal window (this happens by default unless you use the \fB-pid\fR option
of these programs to control a different signal window), the other signal
windows are kept synchronized with the master window.
.PP
Note that you cannot \fIincrease\fR the number of signal windows in a group
once you have started a \fIwave\fR process group, although you can run more
than one process group at a time if you wish.

.SH MENU FILE
.PP
\fIwave\fR uses a simple menu file to allow you to set up analysis
options.  Each line in the file corresponds to a button in the Analyze window
(except for empty lines and lines that begin with `#', which are ignored).
Within each line, the syntax is \fIlabel\fR<tab>\fIaction\fR, where <tab> is
one or more tab characters.  The \fIlabel\fR field is used to identify a
command button in the Analyze window, and the \fIaction\fR field is any command
acceptable to your shell.  \fIbutton-label\fR and \fIaction\fR may include
spaces if needed; if necessary, a `\\' may be used at the end of a line to
indicate that it is continued on the next line.  Before the command is
executed, \fIwave\fR replaces certain tokens with appropriate strings;  these
include:
.TP
\fB$RECORD\fR
The name of the current record.
.TP
\fB$ANNOTATOR\fR
The name of the current input annotator.
.TP
\fB$START\fR
The currently selected `start analysis' time.
.TP
\fB$END\fR
The currently selected `end analysis' time.
.TP
\fB$DURATION\fR
The time interval between \fB$END\fR and \fB$START\fR.
.TP
\fB$LEFT\fR
The time corresponding to the left edge of the signal window.
.TP
\fB$RIGHT\fR
The time corresponding to the right edge of the signal window.
.TP
\fB$WIDTH\fR

wave/wave.c
/* file: wave.c		G. Moody	27 April 1990
			Last revised:	10 June 2005
main() function for WAVE

-------------------------------------------------------------------------------
WAVE: Waveform analyzer, viewer, and editor
Copyright (C) 1990-2005 George B. Moody

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, see <http://www.gnu.org/licenses/>.

You may contact the author by e-mail (george@mit.edu) or postal mail
(MIT Room E25-505A, Cambridge, MA 02139 USA).  For updates to this software,
please visit PhysioNet (http://www.physionet.org/).
_______________________________________________________________________________

*/

#include "wave.h"
#include "xvwave.h"
#include <unistd.h>		/* for getpid declaration */

/* Spot help files for WAVE are located in HELPDIR/wave.  Their names are
   of the form `HELPDIR/wave/*.info'. */
#ifndef HELPDIR
#define HELPDIR		"/usr/local/help"
#endif

main(argc, argv)
int argc;
char *argv[];
{
    char *wfdbp, *hp, *p, *start_string = NULL, *tp, *getenv();
    int do_demo = 0, i, j, mode = 0;
    static char *helppath;
    static int wave_procno;
    void set_frame_footer();

    /* Extract program name for use in error messages. */
    for (p = pname = argv[0]; *p; p++)
	if (*p == '/') pname = p+1;

    /* Initialize non-zero global variables. */
    begin_analysis_time = end_analysis_time = -1L;
    tsa_index = vsa_index = ann_mode = overlap = sig_mode = time_mode =
	grid_mode = -1;
    tscale = 1.0;

    if ((helpdir = getenv("HELPDIR")) == NULL) helpdir = HELPDIR;

    /* Find the record name argument. */
    for (i = 1; i < argc-1; i++) {
	if (strcmp(argv[i], "-r") == 0)
	    break;
    }

    /* Provide (non-window based) usage information if needed. */
    if (i == argc-1 || argc == 1 || getenv("DISPLAY") == NULL)
	help();	/* print info and quit */

    /* Set the path for XView spot help. */
    if ((hp = getenv("HELPPATH")) == NULL)
	hp = "/usr/lib/help";
    helppath = (char *)malloc(strlen(hp) + strlen(helpdir) + 16);
    	/* (strlen("HELPPATH=") + strlen(":") + strlen("/wave") + 1 = 16) */
    if (helppath) {
	sprintf(helppath, "HELPPATH=%s:%s/wave", hp, helpdir);
	putenv(helppath);
    }

    /* Check for requests to open more than one record. */
    strcpy(record, argv[++i]);
    while (p = strchr(record, '+')) {
	static int wave_pid;

	wave_pid = (int)getpid();	/* save the PID of this process */
	if (fork() == 0) {		/* child */
	    wave_ppid = wave_pid;
	    wave_procno++;
	    p++; tp = record;
	    while (*tp++ = *p++)
		;	/* delete the first record name from the string */
	    strcpy(argv[i], record);
	    if (strchr(record, '+') == NULL)
		make_sync_button = 1;
	}
	else {	/* parent */
	    *p = '\0';	/* delete everything after the first record from the
			   string */
	    strcpy(argv[i], record);
	}
    }

    /* Remove any command-line arguments that are specific to another WAVE
       process, and remove the `+n/' prefix from any that are specific to
       this process. */
    for (i = 1; i < argc; i++) {
	if (*argv[i] == '+') {
	    if ((atoi(argv[i]+1) == wave_procno) && (p = strchr(argv[i], '/')))
	      argv[i] = p+1;
	    else {
		int j;
		for (j = i+1; j < argc; j++)
		    argv[j-1] = argv[j];
		argc--;
	    }
	}
    }

    /* Process window system related command-line arguments. */
    strip_x_args(&argc, argv);

    /* Process WAVE-specific arguments. */
    for (i = 1; i < argc; i++) {
	if (*argv[i] == '-')
	    switch (*(argv[i] + 1)) {
	      case 'a':	/* annotator name */
		if (++i >= argc) {
		    fprintf(stderr, "%s: annotator name must follow -a\n",
			    pname);
		    exit(1);
		}
		if (strlen(argv[i]) > ANLMAX) {
		    fprintf(stderr, "%s: annotator name is too long\n",
			    pname);
		    exit(1);
		}
		strcpy(annotator, argv[i]);
		break;
	      case 'd':	/* display resolution */
		if (strcmp(argv[i], "-dpi")) break;
		if (++i >= argc) {
		    fprintf(stderr, "%s: resolution must follow -dpi\n",
			    pname);
		    exit(1);
		}
		(void)sscanf(argv[i], "%lfx%lf", &dpmmx, &dpmmy);
		if (dpmmx < 0.) dpmmx = 0.;
		if (dpmmy <= 0.) dpmmy = dpmmx;
		dpmmx /= 25.4; dpmmy /= 25.4;
		break;
	      case 'D': /* demo mode */
		if (++i >= argc) {
		    fprintf(stderr, "%s: log file name must follow -D\n",
			    pname);
		    exit(1);
		}
		if (strlen(argv[i]) > LNLMAX) {
		    fprintf(stderr, "%s: log file name is too long\n", pname);
		    exit(1);
		}
		strcpy(log_file_name, argv[i]);
		do_demo = 1;
		break;
	      case 'f':	/* start at time specified in following argument */
		if (++i >= argc) {
		    fprintf(stderr, "%s: start time must follow -f\n",
			    pname);
		    exit(1);
		}
		start_string = argv[i];
		if (start_string[0] == '[')
		    time_mode = 1;	/* display absolute times */
		else if (start_string[0] == 's')
		    time_mode = 2;	/* display times in sample intervals */
		break;
	      case 'g':	/* use shades of grey, even on color screen */
		mode |= MODE_GREY;
		break;
	      case 'H':	/* use getvec's high-resolution mode for
			   multi-frequency records */
	        setgvmode(WFDB_HIGHRES);
		break;
	      case 'm':	/* run in monochrome, even on color screen */
		mode |= MODE_MONO;
		break;
	      case 'O':	/* use overlay mode */
		mode |= MODE_OVERLAY;
		break;
	      case 'p': /* add argument to WFDB path */
		if (++i >= argc) {
		    fprintf(stderr,
			    "%s: input file location(s) must follow -p\n",
			    pname);
		    exit(1);
		}
		wfdb_addtopath(argv[i]);
		break;
	      case 'r':	/* record name */
		if (++i >= argc) {
		    fprintf(stderr, "%s: record name must follow -r\n",
			    pname);
		    exit(1);
		}
		if (strlen(argv[i]) > RNLMAX) {
		    fprintf(stderr, "%s: record name is too long\n",
			    pname);
		    exit(1);
		}
		strcpy(record, argv[i]);
		break;
	      case 's':	/* signal list follows */
	        /* count the number of signals */
		for (j = 0; ++i < argc && *argv[i] != '-'; j++)
		    ;
		if (j == 0) {
		    (void)fprintf(stderr,
			     "%s: one or more signal numbers must follow -s\n",
				  pname);
		    exit(1);
		}
		/* allocate storage for the signal list */
		if (siglistlen + j > maxsiglistlen) {
		    if ((siglist = realloc(siglist,
				      (siglistlen+j) * sizeof(int))) == NULL ||
			(base = realloc(base,
				      (siglistlen+j) * sizeof(int))) == NULL ||
			(level = realloc(level,
				 (siglistlen+j) * sizeof(XSegment))) == NULL) {
			(void)fprintf(stderr, "%s: insufficient memory\n",
				      pname);
			exit(2);
		    }
		    maxsiglistlen = siglistlen + j;
		}
		/* fill the signal list */
		for (i -= j; i < argc && *argv[i] != '-'; )
		    siglist[siglistlen++] = atoi(argv[i++]);
		i--;
		sig_mode = 1;	/* display listed signals only (may be
				   overridden using -VS 0) */
		break;
	      case 'S':   /* use shared color map only */
		mode |= MODE_SHARED;
		break;
	      case 'V':	/* View options */
		switch (*(argv[i]+2)) {
		  case 'a':	/* display annotation `aux' fields */
		    show_aux = 1;
		    break;
		  case 'A':	/* set annotation display mode */
		    if (++i >= argc) {
			fprintf(stderr,
			       "%s: annotation display mode must follow -VA\n",
				pname);
			exit(1);
		    }
		    ann_mode = atoi(argv[i]);
		    break;
		  case 'b':	/* display signal baselines */
		    show_baseline = 1;
		    break;
		  case 'c':	/* display annotation `chan' fields */
		    show_chan = 1;
		    break;
		  case 'G':	/* set grid display mode */
		    if (++i >= argc) {
			fprintf(stderr,
			       "%s: grid display mode must follow -VG\n",
				pname);
			exit(1);
		    }
		    grid_mode = coarse_grid_mode = fine_grid_mode =
			atoi(argv[i]);
		    break;
		  case 'l':	/* display cursor levels */
		    show_level = 1;
		    break;
		  case 'm':	/* display annotation markers */
		    show_marker = 1;
		    break;
		  case 'n':	/* display annotation `num' fields */
		    show_num = 1;
		    break;
		  case 'N':	/* display signal names */
		    show_signame = 1;
		    break;
		  case 's':	/* display annotation subtypes */
		    show_subtype = 1;
		    break;
		  case 'S':	/* set signal display mode */
		    if (++i >= argc) {
			fprintf(stderr,
				"%s: signal display mode must follow -VS\n",
				pname);
			exit(1);
		    }
		    sig_mode = atoi(argv[i]);
		    break;
		  case 't':	/* set time scale array index */
		    if (++i >= argc) {
			fprintf(stderr,
				"%s: time scale choice must follow -Vt\n",
				pname);
			exit(1);
		    }
		    tsa_index = coarse_tsa_index = fine_tsa_index =
			atoi(argv[i]);
		    break;
		  case 'T':	/* set time display mode */
		    if (++i >= argc) {
			fprintf(stderr,
			       "%s: time display mode must follow -VT\n",
				pname);
			exit(1);
		    }
		    time_mode = atoi(argv[i]);
		    break;
		  case 'v':	/* set amplitude scale array index */
		    if (++i >= argc) {
			fprintf(stderr,
				"%s: amplitude scale choice must follow -Vv\n",
				pname);
			exit(1);
		    }
		    vsa_index = atoi(argv[i]);
		    break;
		}
		break;
	    }
    }

    /* Provide help if no record was specified. */
    if (record[0] == '\0') help();

    /* Set up base frame, quit if unsuccessful. */
    if (initialize_graphics(mode)) exit(1);

    /* Make sure that the WFDB path begins with an empty component (otherwise,
       annotation editing may become confused). */
    wfdbp = getwfdb();
    if (*wfdbp != ':') {
	char *nwfdbp;

	if ((nwfdbp = (char *)malloc(strlen(wfdbp)+2)) == NULL) {
	    fprintf(stderr, "%s: memory allocation error\n", pname);
	    exit(1);
	}
	sprintf(nwfdbp, ":%s", wfdbp);
	setwfdb(nwfdbp);
    }

    /* Make sure there is a calibration database defined. */
    if (!getenv("WFDBCAL"))
	putenv("WFDBCAL=wfdbcal");
    /* Initialize the annotation table. */
    (void)read_anntab();

    /* Open the selected record (and annotation file, if specified). */
    if (record_init(record)) {
	if (annotator[0]) {
	    af.name = annotator; af.stat = WFDB_READ;
	    nann = 1;
	    annot_init();
	}
    }

    /* Indicate that the input annotation file must be saved before writing
       any edits. */
    savebackup = 1;

    /* If requested, set the start time. */
    if (start_string) {
	display_start_time = strtim(start_string);
	if (start_string[0] == '[') {
	    if (display_start_time > 0L)
		/* a time before the beginning of the record was requested --
		   go to the beginning of the record instead */
		display_start_time = 0;
	    else
		/* strtim converts a valid absolute time string into a negated
		   sample number -- change it to a positive sample number */
		display_start_time = -display_start_time;
	}
    }
    set_frame_footer();

    /* If requested, start demonstration. */
    if (do_demo)
      start_demo();

    /* Do initial display */
    XFillRectangle(display, osb, clear_all, 0, 0, canvas_width+mmx(10), canvas_height);
    do_disp();

    /* Enter the main loop. */
    display_and_process_events();

    exit(0);
}
